src/core/thread/anycast_locator.hpp

```diff
@@ -0,0 +1,114 @@
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes definitions for Anycast Locator functionality.
+ */
+
+#ifndef ANYCAST_LOCATOR_HPP_
+#define ANYCAST_LOCATOR_HPP_
+
+#include "openthread-core-config.h"
+
+#if OPENTHREAD_CONFIG_TMF_ANYCAST_LOCATOR_ENABLE
+
+#include "coap/coap.hpp"
+#include "common/locator.hpp"
+#include "common/non_copyable.hpp"
+#include "net/ip6_address.hpp"
+
+namespace ot {
+
+/**
+ * This class implement Anycast Locator functionality which allows caller to determine the mesh local EID and RLOC16 of
+ * the closest destination of an anycast address (if any).
+ *
+ * The closest destination is determined based on the the current routing table and path costs within the Thread mesh.
+ *
+ */
+class AnycastLocator : public InstanceLocator, private NonCopyable
+{
+public:
+    /**
+     * This function pointer type defines the callback to notify the outcome of a request.
+     *
+     */
+    typedef otThreadAnycastLocatorCallback Callback;
+
+    /**
+     * This constructor initializes the `AnycastLocator` object.
+     *
+     * @param[in]  aInstance  A reference to the OpenThread instance.
+     *
+     */
+    explicit AnycastLocator(Instance &aInstance);
+
+    /**
+     * This method requests the closest destination of a given anycast address to be located.
+     *
+     * If a previous `Locate()` request is ongoing, a subsequent call to this method will cancel and replace the
+     * earlier request.
+     *
+     * @param[in] aAnycastAddress   The anycast address to locate the closest destination of.
+     * @param[in] aCallback         The callback handler to report the result.
+     * @param[in] aContext          An arbitrary context used with @p aCallback.
+     *
+     * @retval kErrorNone         The request started successfully. @p aCallback will be invoked to report the result.
+     * @retval kErrorNoBufs       Out of buffer to prepare and send the request message.
+     * @retval kErrorInvalidArgs  The @p aAnycastAddress is not a valid anycast address or @p aCallback is `nullptr`.
+     *
+     */
+    Error Locate(const Ip6::Address &aAnycastAddress, Callback aCallback, void *aContext);
+
+    /**
+     * This method indicates whether an earlier request is ongoing.
+     *
+     * @returns TRUE if a previous request is ongoing, FALSE otherwise.
+     *
+     */
+    bool IsRunning(void) const { return (mCallback != nullptr); }
```

Comment:
```suggestion
    bool IsBusy(void) const { return (mCallback != nullptr); }
```

Reply:
BTW, this method not used at all.

Reply:
I prefer `IsRunning()` (already used the same name in a bunch of other places).
yes, it is not used currently (it's a simple inline getter).

Reply:
I actually renamed it to `IsInProgress()`